### DISCUSSION TITLE
It would be great if this project had some kind of style consistency.

### DISCUSSION THREAD

**Discussion by Inmortak** [sympy/sympy#26190] [General]
I find it kind of messy the lack of convention in names classes, functions, objects, etc. 

It would be great if this project applied the python style guide: 

https://peps.python.org/pep-0008/

**Comment by oscarbenjamin** [comment]
Broadly PEP 8 is followed.

Note that if you read the PEP carefully it does not say that all code should be rewritten in any uniform style.

Otherwise be more specific because the description above is too vague to be actionable.

**Reply by Inmortak** [reply]
Yeah, I am referring specifically to function/method names, which are supposed to be lowercase with words separated by underscores. 

We have, many examples that violate this rule. And I find myself frequently having to go back to the documentation to verify which is the syntax of  each case:

tomatrix()
tolist() 
applyfunc()
tensorcontraction()
Predicate()
Partition()
IndexedBase
apply_finite_diff()
differentiate_finite()

As you can see, pretty much all possible conventions are used, depending on the module.

**Reply by oscarbenjamin** [reply]
Most names match what pep8 would suggest. If you propose a change then flesh it out more fully. Which names should change and to what?

**Reply by asmeurer** [reply]
Underscores vs. non-underscores are hard. It's pretty common to omit an underscore for two letter words like `to` or `as`. I agree that something like `tensorcontraction` should probably have an underscore. Unfortunately, changing names like this requires a deprecation, which is annoying. So if we want to do a cleanup like this it would be best to do a bunch of them at once, or else just leave things as they are.

I think most of the names that are capitalized are indeed classes. Although this is also somewhat of a complicated issue for SymPy. Most symbolic functions in SymPy are classes (like `sin`, `cos`, `exp`, ...), but we use lowercase names for them because 1) it's more important to match the mathematical conventions, and 2) the fact that they are classes (rather than instances) is somewhat of an implementation detail, which may change in the future.

To me more annoying is when different words or combinations of words are used for the same thing in different places (`differentiate_finite` vs. `apply_finite_diff` is a good example of this). But again, this is also annoying to change, so if we really did want to clean this up (e.g., maybe for a potential SymPy 2.0), we would want to do a thorough review of the public namespace to try to catch them all. This isn't really a PEP 8 thing, but rather just an issue with different people creating function names without being consistent with what is already there.
